gnomad/resources/grch38/gnomad.py

```diff
@@ -430,6 +430,79 @@ def release_vcf_path(data_type: str, version: str, contig: str) -> str:
     return f"gs://gcp-public-data--gnomad/release/{version}/vcf/{data_type}/gnomad.{data_type}.{version_prefix}{version}.sites{contig}.vcf.bgz"
 
 
+def add_grpMaxFAF95_v3(ht: hl.Table) -> hl.Table:
+    """
+    Add a grpMaxFAF95 annotation using the v3 style faf and faf_index_dict structures.
+
+    :param ht: Input hail table.
+    :return: Annotated hail table.
+    """
+    ht = ht.annotate(
+        grpMaxFAF95=hl.rbind(
+            hl.sorted(
+                hl.array(
+                    [
+                        hl.struct(
+                            faf=ht.faf[ht.faf_index_dict[f"{pop}-adj"]].faf95,
+                            population=pop,
+                        )
+                        for pop in FAF_POPS
+                    ]
+                ),
+                key=lambda f: (-f.faf, f.population),
+            ),
+            lambda fafs: hl.if_else(
+                hl.len(fafs) > 0,
+                hl.struct(
+                    popmax=fafs[0].faf,
+                    popmax_population=hl.if_else(
+                        fafs[0].faf == 0, hl.missing(hl.tstr), fafs[0].population
+                    ),
+                ),
+                hl.struct(
+                    popmax=hl.missing(hl.tfloat), popmax_population=hl.missing(hl.tstr)
```

Comment:
Remind me, the I'm having a hard time parsing the Schema. 'GrpMaxFAF95' is required to exist but can just return values of None and None within it?

Reply:
It's not required to exist anymore, instead of making it required and allowing None, they just made it optional.

https://github.com/ga4gh/va-spec/blob/1.0-alpha/schema/cohortAlleleFreq.json#L341-L359

Reply:
The way I'm always reading it, in this code, it looks like GrpMaxFAF95 always exists but everything it's populated with is allowed to be None. 1) Am I reading this implementation right 2) this implementation is allowed, right? Sorry if this is v base level

Reply:
ah okay, makes sense ! thanks for the answer